### PR TITLE
configure public key and related license variables at build time

### PR DESCRIPTION
**Pull request by neallred**
#### Summary

**NOTE: This PR's test public key file is the same as the production public key file, and will be merged like this to testing that the pipelines works as expected. This will also make for the smallest possible diff (and easiest revert, if needed) when we make the real test license key change.**

Add test environment public key and other configurations (like where to request trials and purchase licenses from), and choose between them at build time according to the `testpublickey` go build tag. A later PR will change the actual values, this PR just sets up the structure to easily change those values in a future PR.

How licensing currently works, all licenses files are a base64 encoded json blob, with a signature attached to the end of the blob that verifies the json blob's contents and the authenticity that Mattermost generated the license. When licenses are applied, we validate them by checking against the public key stored in this repo.

We are currently always hardcoded to use the production environment license-generator's public key, including in our test environments, which is not best practice, and has made QA-ing some scenarios difficult. It has become more of a problem recently since we added the ability to purchase a license from within the Mattermost application. We have a series of work we'd like to do so that test environments and local development use test licenses instead of production licenses. Here's a link to the [epic](https://mattermost.atlassian.net/browse/MM-50159) and the [spec doc](https://docs.google.com/document/d/1jCEysaZanvZI8grSyjjKZQW10ZErCeXgkDW7lnwB3Uo) outlining that work.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-50177

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by noxer** [comment]
I'm not sure what this change is good for. Could you explain why it's necessary, @neallred?

**Comment by neallred** [comment]
Sure! How licensing currently works, all licenses files are a base64 encoded json blob, with a signature attached to the end of the blob that verifies the json blob's contents and the authenticity that Mattermost generated the license. When licenses are applied, we validate them by checking against the public key here.

This is the production environment license-generator's public key, and it is currently used in all our test environments as well, which is not best practice, and has made QA-ing some scenarios difficult from the beginning. Its become more a problem recently since we added the ability to purchase a license from within the Mattermost application. We have a series of work we'd like to do so that test environments and local development uses test licenses instead of production licenses. Here's a link to the [epic](https://mattermost.atlassian.net/browse/MM-50159) and the [spec doc](https://docs.google.com/document/d/1jCEysaZanvZI8grSyjjKZQW10ZErCeXgkDW7lnwB3Uo) outlining that work.

This PR will enable *additional future PRs* to change the local development process so that it seamlessly uses the environment test public key for those with the mattermost/enterprise repo (those developing locally without mattermost/enterprise tend to not need to do anything with licenses because they don't have any of the enterprise code requiring a license).

**Comment by neallred** [comment]
@noxer @marianunez gentle ping for review

**Review by marianunez — APPROVED** [review]

File: app/platform/license.go [hunk 1]
```diff
@@ -24,13 +24,20 @@ import (
 const (
 	LicenseEnv                = "MM_LICENSE"
 	LicenseRenewalURL         = "https://customers.mattermost.com/subscribe/renew"
+	LicenseRenewalURLTest     = "https://portal.test.cloud.mattermost.com/subscribe/renew"
```
Comment (marianunez) [inline app/platform/license.go:27]: Wondering if we should use `CWSURL` config setting here as the base url and not have to keep these two separate urls?

File: app/platform/license.go [hunk 2]
```diff
@@ -24,13 +24,20 @@ import (
...
 	JWTDefaultTokenExpiration = 7 * 24 * time.Hour // 7 days of expiration
 )
 
 var (
 	RequestTrialURL = "https://customers.mattermost.com/api/v1/trials"
 )
 
+func init() {
+	if !model.IsProdLicensePublicKey {
+		RequestTrialURL = "https://portal.test.cloud.mattermost.com/api/v1/trials"
```
Comment (marianunez) [inline app/platform/license.go:37]: Same as above

File: app/platform/license.go (hunk 1 above)
Comment (neallred) [reply app/platform/license.go:27]: Yes, I think that is a good idea.

File: model/config.go [hunk 3]
```diff
@@ -239,9 +239,12 @@ const (
 	Office365SettingsDefaultTokenEndpoint   = "https://login.microsoftonline.com/common/oauth2/v2.0/token"
 	Office365SettingsDefaultUserAPIEndpoint = "https://graph.microsoft.com/v1.0/me"
 
-	CloudSettingsDefaultCwsURL    = "https://customers.mattermost.com"
-	CloudSettingsDefaultCwsAPIURL = "https://portal.internal.prod.cloud.mattermost.com"
-	OpenidSettingsDefaultScope    = "profile openid email"
+	CloudSettingsDefaultCwsURL        = "https://customers.mattermost.com"
+	CloudSettingsDefaultCwsAPIURL     = "https://portal.internal.prod.cloud.mattermost.com"
+	CloudSettingsDefaultCwsURLTest    = "https://portal.test.cloud.mattermost.com"
+	CloudSettingsDefaultCwsAPIURLTest = "https://portal.test.cloud.mattermost.com"
```
Comment (marianunez) [inline model/config.go:245]: Is it intended that the API test one is the same? I think we do have an internal test endpoint

File: Makefile [hunk 4]
```diff
@@ -41,6 +45,7 @@ endif
 
 ifeq ($(BUILD_NUMBER),dev)
 	export MM_FEATUREFLAGS_GRAPHQL = true
+	GOTAGS += "testlicensekey"
```
Comment (marianunez) [inline Makefile:48]: Is this the case for spinwicks and cloud test servers ? They may have a build number set

File: model/config.go (hunk 3 above)
Comment (neallred) [reply model/config.go:245]: It is intentional. [Last someone tried using the other test endpoint](https://community.mattermost.com/private-core/pl/h4gxtmjkgbgcimez3hegs1y7ah), it didn't seem to work so we changed the cws docs to recommend using the same: https://github.com/mattermost/customer-web-server/pull/1086/files

Comment (marianunez) [reply model/config.go:245]: Ah in the context of local dev environment but in the test environment I believe it does need to point to the internal endpoint `https://api.internal.test.cloud.mattermost.com`

File: Makefile (hunk 4 above)
Comment (neallred) [reply Makefile:48]: build number is only `dev` for local development. A separate change in [build-ee](https://git.internal.mattermost.com/mattermost/ci/build-ee/-/tree/MM-50182) will control the mm-ee-test image

File: model/config.go (hunk 3 above)
Comment (neallred) [reply model/config.go:245]: good catch, will update this value

File: Makefile (hunk 4 above)
Comment (marianunez) [reply Makefile:48]: Gotcha! That makes sense 👍

File: model/config.go (hunk 3 above)
Comment (neallred) [reply model/config.go]: updated

File: app/platform/license.go (hunk 1 above)
Comment (neallred) [reply app/platform/license.go]: changed

**Review by marianunez — APPROVED** [review]
LGTM. Added @jupenur for a second set of eyes from security perspective, Juho feel free to change as needed

**Comment by isacikgoz** [comment]
@neallred overall this looks good to me but I couldn't understand why yo didn't use the ldflags? Is it only for being able to use `const`?

**Review by jupenur — APPROVED** [review]
LGTM

**Comment by neallred** [comment]
@isacikgoz yep, pretty much; we want to be able to set this at build time instead of run time. This PR originally did use ldflags, and was going to rely on CI build setting the public key. This also required representing the public key as a base64 string, which isn't a big deal but is a little less clear.

Tim realized we could simplify the build setup by storing both public keys as files and then embedding the right one. I like that approach, but can't think of a way to achieve it without using go tags. We could do something like put it all in an `init` function, but then that happens at run time :shrug: 

Can you think of a more elegant solution? Does this approach seem ok?

**Review by isacikgoz — APPROVED** [review]
> Can you think of a more elegant solution? Does this approach seem ok?

This approach seem to be okay, I just wanted to dig in between difference of using tags and flags. For embedding I think I got your point, this looks good to me!

**Pull request merged**